File: fixtures/test-repo/app/models/user_model.rb
# User model that extends BaseModel
require_relative 'base_model'

class UserModel < BaseModel
  attr_accessor :username, :email, :first_name, :last_name, :active

  def initialize(attributes = {})
    super(attributes)
    @username = attributes[:username]
    @email = attributes[:email]
    @first_name = attributes[:first_name]
    @last_name = attributes[:last_name]
    @active = attributes.fetch(:active, true)
  end

  def self.find_by_username(username)
    storage.find { |user| user.username == username }
  end

  def self.find_by_email(email)
    storage.find { |user| user.email == email }
  end

  def self.active_users
    where(active: true)
  end

  def self.inactive_users
    where(active: false)
  end

  def full_name
    "#{@first_name} #{@last_name}".strip
  end

  def display_name
    full_name.empty? ? @username : full_name
  end

  def activate!
    update(active: true)
  end

  def deactivate!
    update(active: false)
  end

  def change_email(new_email)
    raise ArgumentError, "Email already taken" if self.class.find_by_email(new_email)
    update(email: new_email)
  end

  def change_username(new_username)
    raise ArgumentError, "Username already taken" if self.class.find_by_username(new_username)
    update(username: new_username)
  end

  def to_h
    super.merge({
      username: @username,
      email: @email,
      first_name: @first_name,
      last_name: @last_name,
      active: @active,
      full_name: full_name,
      display_name: display_name
    })
  end

  # Validation methods
  def valid?
    validate_username && validate_email
  end

  def errors
    @errors ||= []
  end

  private

  def validate_username
    if @username.nil? || @username.strip.empty?
      errors << "Username cannot be empty"
      return false
    end
    
    if @username.length < 3
      errors << "Username must be at least 3 characters"
      return false
    end
    
    true
  end

  def validate_email
    if @email.nil? || @email.strip.empty?
      errors << "Email cannot be empty"
      return false
    end
    
    unless @email.match?(/\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i)
      errors << "Email format is invalid"
      return false
    end
    
    true
  end
end 

File: fixtures/typescript/test-repo/app/models/user_model.ts
import { BaseModel } from './base_model';

export class UserModel extends BaseModel {
  public username?: string;
  public email?: string;
  public firstName?: string;
  public lastName?: string;
  public active: boolean;
  private errors: string[] = [];

  constructor(attributes: Record<string, any> = {}) {
    super(attributes);
    this.username = attributes.username;
    this.email = attributes.email;
    this.firstName = attributes.firstName;
    this.lastName = attributes.lastName;
    this.active = attributes.active !== undefined ? attributes.active : true;
  }

  static findByUsername(username: string): UserModel | undefined {
    return this.all().find((user: UserModel) => user.username === username);
  }

  static findByEmail(email: string): UserModel | undefined {
    return this.all().find((user: UserModel) => user.email === email);
  }

  static activeUsers(): UserModel[] {
    return this.where({ active: true });
  }

  static inactiveUsers(): UserModel[] {
    return this.where({ active: false });
  }

  get fullName(): string {
    return `${this.firstName || ''} ${this.lastName || ''}`.trim();
  }

  get displayName(): string {
    return this.fullName || this.username || '';
  }

  activate(): this {
    return this.update({ active: true });
  }

  deactivate(): this {
    return this.update({ active: false });
  }

  changeEmail(newEmail: string): this {
    if (UserModel.findByEmail(newEmail)) {
      throw new Error("Email already taken");
    }
    return this.update({ email: newEmail });
  }

  changeUsername(newUsername: string): this {
    if (UserModel.findByUsername(newUsername)) {
      throw new Error("Username already taken");
    }
    return this.update({ username: newUsername });
  }

  toObject(): Record<string, any> {
    return {
      ...super.toObject(),
      username: this.username,
      email: this.email,
      firstName: this.firstName,
      lastName: this.lastName,
      active: this.active,
      fullName: this.fullName,
      displayName: this.displayName
    };
  }

  valid(): boolean {
    this.errors = [];
    return this.validateUsername() && this.validateEmail();
  }

  getErrors(): string[] {
    return [...this.errors];
  }

  private validateUsername(): boolean {
    if (!this.username || this.username.trim().length === 0) {
      this.errors.push("Username cannot be empty");
      return false;
    }
    
    if (this.username.length < 3) {
      this.errors.push("Username must be at least 3 characters");
      return false;
    }
    
    return true;
  }

  private validateEmail(): boolean {
    if (!this.email || this.email.trim().length === 0) {
      this.errors.push("Email cannot be empty");
      return false;
    }
    
    const emailRegex = /^[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+$/i;
    if (!emailRegex.test(this.email)) {
      this.errors.push("Email format is invalid");
      return false;
    }
    
    return true;
  }
}
